### ISSUE TITLE
Any official plan to add support for Windows?

### ISSUE THREAD

**Issue by diman82** [horovod/horovod#2591]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.

Any time estimation when Windows support will be available?

**Comment by tgaddair** [comment]
Hey @diman82, it's not on our roadmap at the moment, but this would definitely be a great addition.  What kind of setup are you looking to run this on?  Single machine with multiple GPUs, private cluster, or something else?

**Comment by diman82** [comment]
Hi @tgaddair. First of all, I'm looking for a development setup - I work on a windows workstation (with GPU), and I would like to test development on it.
For production needs, my personal need is running a private cluster (multiple windows workstations with single GPU).

**Comment by tgaddair** [comment]
Thanks @diman82, that makes sense.  The biggest question I would have here is: what's the state of NCCL support for Windows?  Assuming that's available, it should be doable.  The other question is whether it makes the most sense to do this as part of Windows directly, or through the Linux subsystem for Windows.  I'm inclined to think the Linux subsystem is the best route, but would be good to get your thoughts on this as someone working on Windows.

**Comment by diman82** [comment]
@tgaddair NCCL is not planned to be added to vanilla Windows, and yes - best shot is to progress with WSL2 development.
But I would like to tackle this question from another angle: NCCL is basically for supporting multi-GPU runtime, right?
As far as I understand, it's not needed for a single-GPU operation, so.. Is it possible to add horovod windows support without NCCL? Or is it considered to be a big rewrite of the codebase?

**Comment by tgaddair** [comment]
Hey @diman82, Horovod does not need NCCL, but if you're only running with a single GPU, then Horovod will not do anything.  For the situation you describe, the best route may be to add a "stub" mode for Horovod where we don't call any native libraries that only works with a single process.  Would something like that work for you?

**Comment by diman82** [comment]
@tgaddair Maybe I'm missing something - doesn't Horovod support multi-node cluster (several nodes running single GPU) as well? This is an actual setup I have.
A stub mode can be beneficial for development/creating mocked tests..

**Comment by tgaddair** [comment]
@diman82 Horovod does indeed support multi-node training.  I assumed when you said single-GPU, you meant 1 GPU in total.

For multi-node, you would still want to use NCCL in most cases (assuming your GPU is reasonably performant).  To better understand your scenario: are you imagining running a training job with multiple single-GPU Windows machines?  Is that necessary for your development environment, or would running with single-machine / single-GPU suffice?

**Comment by diman82** [comment]
@tgaddair Yes, I would like to have a training job running with multiple single-GPU Windows machines,as it's  necessary for my development/production environment.
Is there any point running a single-machine training job with horovod, besides having a nice wrapper for supported high-level frameworks (like TF/Pytorch/etc..)?

**Comment by tgaddair** [comment]
There's no practical performance benefit to running single-machine Horovod.  The idea would usually be to test training locally before scaling it up to multiple workers without needing to modify the code.

**Comment by diman82** [comment]
@tgaddair Totally agree - single-machine Horovod is needed only for testing purposes.
